### PR TITLE
Pv2 1627 Learner withdrawn at end of previous academic year after payments made in current AY, clawback of payments required

### PR DESCRIPTION
**Pull request by CarlHA**
Only Price episodes which are not in the current academic year have been excluded. Learning deliveries for previous AYs have not been excluded and if these are still an issue they will be addressed in another ticket.

### CODE REVIEW

File: src/SFA.DAS.Payments.AcceptanceTests.Core/AimPeriodMatcher.cs [hunk 1]
```diff
@@ -16,19 +19,42 @@ public static class AimPeriodMatcher
...
+            if (Array.Exists(durationElements, x => x.Contains("month")))
             {
-                var months = int.Parse(duration.Replace("months", string.Empty));
+                var months = int.Parse(durationElements[0]);
+
+                if (durationElements.Contains("-") && Array.Exists(durationElements, x => x.Contains("day")))
+                {
+                    // for now assume the number of days is irrelevant as it should never go back more than 1 month, so no calc required.
+                    months -= 1;
+                }
```
Comment (CarlHA) [inline src/SFA.DAS.Payments.AcceptanceTests.Core/AimPeriodMatcher.cs]: The above logic has been taken from the PDI testing to allow durations such as 7 months - 1 day to be used in the feature file to give precise end dates.

File: src/SFA.DAS.Payments.AcceptanceTests.Core/Extensions.cs [hunk 2]
```diff
@@ -45,16 +47,23 @@ public static decimal ToPercent(this string stringPercent)
...
-                var months = int.Parse(duration.Replace("month", string.Empty));
-                return startDate.ToDate().AddMonths(months) - startDate.ToDate();
+                if (durationElements.Contains("-") && Array.Exists(durationElements, x => x.Contains("day")))
+                {
+                    return startDate.ToDate().AddMonths(int.Parse(durationElements[0])).
+                               AddDays(int.Parse(durationElements[Array.FindIndex(durationElements, 0, x => x.Contains("-")) + 1]) * -1) - startDate.ToDate();
+                }
+
+                return startDate.ToDate().AddMonths(int.Parse(durationElements[0])) - startDate.ToDate();
             }
```
Comment (CarlHA) [inline src/SFA.DAS.Payments.AcceptanceTests.Core/Extensions.cs]: Added for the same reason as the code in the AimPeriodMatcher

File: src/SFA.DAS.Payments.AcceptanceTests.EndToEnd/Tests/Inconsistent Submission Data/PV2-1627 Learner Removed at end of AY in R14.feature [hunk 3]
```diff
@@ -0,0 +1,107 @@
+﻿@ignore
```
Comment (CarlHA) [inline src/SFA.DAS.Payments.AcceptanceTests.EndToEnd/Tests/Inconsistent Submission Data/PV2-1627 Learner Removed at end of AY in R14.feature]: This feature file replicates the scenario but has been tagged as ignore as changes to the test framework to get this to run breaks most of the other tests.  Those framework changes have been removed.

File: src/SFA.DAS.Payments.EarningEvents.Application.UnitTests/SubmittedLearnerAimBuilderTest.cs [hunk 4]
```diff
@@ -144,5 +144,100 @@ public void TestAimsAreCreatedFromCommand()
             mathsAndEnglishAim.Ukprn.Should().Be(processLearnerCommand.Ukprn);
 
         }
+
+        [Test]
+        public void PriorAcademicYearAimsAreExcluded()
+        {
+            var plc = new ProcessLearnerCommand
```
Comment (benagain) [inline src/SFA.DAS.Payments.EarningEvents.Application.UnitTests/SubmittedLearnerAimBuilderTest.cs]: Would it be possible to put this inside the `LearnerBuilder`?

File: src/SFA.DAS.Payments.AcceptanceTests.Core/Extensions.cs [hunk 5]
```diff
@@ -45,16 +47,23 @@ public static decimal ToPercent(this string stringPercent)
                 return null;
             }
 
-            if (duration.Contains("months"))
-            {
-                var months = int.Parse(duration.Replace("months", string.Empty));
-                return startDate.ToDate().AddMonths(months) - startDate.ToDate();
-            }
-            else if (duration.Contains("month"))
+            var options = RegexOptions.None;
```
Comment (keenesolutions) [inline src/SFA.DAS.Payments.AcceptanceTests.Core/Extensions.cs]: This code is very similar to the code in AimPeriodMatcher. GetEndPeriodForAim can it be refactored into an extension method do that the core logic can be used by both?

File: src/SFA.DAS.Payments.EarningEvents.Application.UnitTests/SubmittedLearnerAimBuilderTest.cs (hunk 4 above)
Comment (CarlHA) [reply src/SFA.DAS.Payments.EarningEvents.Application.UnitTests/SubmittedLearnerAimBuilderTest.cs]: Yes it is possible but not quickly or easily.  I have refrained due to time constraints on getting this fix tested and deployed.

File: src/SFA.DAS.Payments.AcceptanceTests.Core/Extensions.cs (hunk 5 above)
Comment (CarlHA) [reply src/SFA.DAS.Payments.AcceptanceTests.Core/Extensions.cs]: They are similar but subtly different, that I would not want to waste effort attempting to refactor.

**Pull request merged**